### PR TITLE
Enable students to join after a session has begun

### LINKED ISSUE #432: Allowing students to join session after it has started
Currently once a session starts, nobody can join it. This will not work well in a large lecture setting. Also, the current way of joining a session will not work well. 

- a new URL parameter ?sessionid=434 shall be interpreted to let a user automatically join a session. This will be passed down from the LTI parameter
- we need to make sure that our operators/activities are robust against "late-joiners"
- joining a p1 or a p3 is "easy", joining a p2 is more complicated. currently the runner will stop on 'Loading' if a user joins a p2 without being part of the social group - probably because it is waiting for a reactive data structure that does not exist... We should probably still pass this user to the activity, and let the activity decide how to deal with him/her? 

In the specific case of UNIL, a new user can join during a1 (individual work) or a2 (joining a group), however if they join once a group is constituted, the ideal thing would be that group members see their group ID, and the newcomer will see the "join group activity", and then be able to belatedly join the existing group... With live operators and control operators, this could be possible (let's say the ac-prox is open for the duration of a3, but closes once a student has chosen a group)... But this could be tricky... Worst case, if you join at a p2, we could display "Sorry, you joined to late, please collaborate directly with the person next to you"...

### PR DESCRIPTION
**Pull request by houshuang**
- All new sessions get a slug generated (in add.session). This is displayed on the top bar of the teacher view
- Students can access the session through {FROG_URL}/slug (optionally with ?login=xxx appended)
- System checks if session can still be joined (no p2 activities have so far been active), and if so, adds user to session
- After user has entered a session, the session ID is deleted from the browser location bar. This is to enable join.all from teacher view to still work. The current session is still stored, and will be displayed unless explicitly overridden by student by providing /{slug}.

It's kind of important that the initial login is as efficient as possible, and I think it might still be possible to optimize it - however the first priority is that it always works correctly. 

Closes #432

### CODE REVIEW

File: frog/imports/api/sessions.js [hunk 1]
```diff
@@ -176,6 +187,17 @@ Meteor.methods({
       connections: Connections.find({ graphId }).fetch()
     });
 
+    const slugs = Sessions.find({}, { fields: { slug: 1 } })
+      .fetch()
+      .map(x => x.slug);
+    let slug;
+    while (true) {
```
[the remark this answers is not in the document]
Comment (houshuang) [reply frog/imports/api/sessions.js]: I look forward to the rewrite from the CS PhD so I can learn and improve :)

File: frog/imports/api/sessions.js [hunk 2]
```diff
@@ -176,6 +187,17 @@ Meteor.methods({
       connections: Connections.find({ graphId }).fetch()
     });
 
+    const slugs = Sessions.find({}, { fields: { slug: 1 } })
```
[the remark this answers is not in the document]
Comment (houshuang) [reply frog/imports/api/sessions.js]: It finds all ({}) sessions, and selects only the field slug. The first argument is the search term, and the second the information we want.

File: frog/imports/api/sessions.js [hunk 3]
```diff
@@ -25,12 +28,20 @@ export const setTeacherSession = (sessionId: string) => {
...
-  Meteor.users.update(
-    { _id: Meteor.userId() },
-    { $set: { 'profile.currentSession': sessionId } }
-  );
+  Meteor.call('set.studentsession', sessionId, Meteor.userId());
 };
 
+Meteor.methods({
+  'set.studentsession': (sessionId, studentId) =>
+    Meteor.users.update(
```
[the remark this answers is not in the document]
Comment (houshuang) [reply frog/imports/api/sessions.js]: Somehow I got tons of "access denied" errors when trying to use users.update. I don't know why, and spent some time googling etc, but changing this fixed it so...

[the remark this answers is not in the document]
Comment (lfaucon) [reply frog/imports/api/sessions.js]: Ok, I get it.

Writing on the user collection is restricted. Users can only freely write on the field `profile`. The server has right to write on any field

**Comment by lfaucon** [comment]
@houshuang, I identified two problems that I fixed:
- the `undefined` values for payload created bugs
- The product were generated from reactive data based on the list of instances computed using `object.globalStructure.studentIds` (see `doGetInstances`). But this list was not being updated when new students join.

**Pull request merged**